lambda/jvm/src/main/scala/feral/lambda/IOLambdaPlatform.scala

```diff
@@ -16,53 +16,41 @@
 
 package feral.lambda
 
-import cats.data.OptionT
 import cats.effect.IO
-import cats.effect.kernel.Resource
 import com.amazonaws.services.lambda.{runtime => lambdaRuntime}
-import io.circe
 import io.circe.Printer
+import io.circe.jawn
 import io.circe.syntax._
 
 import java.io.InputStream
 import java.io.OutputStream
 import java.io.OutputStreamWriter
+import java.nio.channels.Channels
 import scala.concurrent.Await
-import scala.concurrent.duration.Duration
+import scala.concurrent.duration._
 
 private[lambda] abstract class IOLambdaPlatform[Event, Result]
     extends lambdaRuntime.RequestStreamHandler { this: IOLambda[Event, Result] =>
 
   final def handleRequest(
       input: InputStream,
       output: OutputStream,
-      context: lambdaRuntime.Context): Unit = {
+      runtimeContext: lambdaRuntime.Context): Unit = {
     val (dispatcher, lambda) =
-      Await.result(setupMemo, Duration.Inf)
+      Await.result(setupMemo, runtimeContext.getRemainingTimeInMillis().millis)
 
-    dispatcher.unsafeRunSync {
-      Resource
-        .eval {
-          for {
-            event <- fs2
-              .io
-              .readInputStream(IO.pure(input), 8192, closeAfterUse = false)
-              .through(circe.fs2.byteStreamParser)
-              .through(circe.fs2.decoder[IO, Event])
-              .head
-              .compile
-              .lastOrError
-            context <- IO(Context.fromJava[IO](context))
-            _ <- OptionT(lambda(event, context)).foreachF { result =>
-              Resource.fromAutoCloseable(IO(new OutputStreamWriter(output))).use { writer =>
-                IO.blocking(Printer.noSpaces.unsafePrintToAppendable(result.asJson, writer))
-              }
-            }
-          } yield ()
-        }
-        .onFinalize(IO.blocking(input.close()) &> IO.blocking(output.close()))
-        .use_
-    }
+    val event = jawn.decodeChannel[Event](Channels.newChannel(input)).fold(throw _, identity(_))
+    val context = Context.fromJava[IO](runtimeContext)
+    dispatcher
+      .unsafeRunTimed(
+        lambda(event, context),
+        runtimeContext.getRemainingTimeInMillis().millis
+      )
+      .foreach { result =>
+        val writer = new OutputStreamWriter(output)
+        Printer.noSpaces.unsafePrintToAppendable(result.asJson, writer)
+        writer.flush()
```

Comment:
Ok, so it's a bit unidiomatic to move this and the event decoding out of `IO`. But here's my thinking ...

> In the unlikely case that this assumption is wrong, blocking on the compute before/after the handler seems pretty harmless anyway, since there is no concurrency.

We already know we have to block the lambda "main" thread while we run `IO` on the CE runtime. So by moving all interaction with `input`/`output` outside of `IO` to this thread, if they _are_ blocking (they're not) then at least we're doing it on this already-blocked thread and not clogging up a compute thread.